### PR TITLE
Async router and functionality added to SMap. Map position coords now stored in URL.

### LINKED ISSUE #1206: SMAP: Redirecting to specific location tab
### Current behavior
Every time the location detail router is called, it displays the overview tab, regardless of which tab you were initially on.

### Expected behavior
I should be able to specify which tab to redirect to. So if I, for example, detach, upload, or add a resource for a location, I would expect to stay on the resource tab, but am instead redirected to the overview tab. Same goes for adding relationships.

### LINKED ISSUE #1207: SMAP: Router should handle permission errors
### Steps to reproduce the error
Try doing anything with a user that doesn't have permission.

### Actual behavior
It dumps all of the HTML for a new page into the detail panel.

### Expected behavior
It should either show or hide the error message accordingly.

### LINKED ISSUE #1310: SMAP: Accessing a location or relationship from url should center on location geometry
### Steps to reproduce the error
- Access a location through a direct URL rather than GUI

### Actual behavior
- Map stays on project extent which is loaded by default

### Expected behavior
- Zoom to current location's geometry.

### LINKED ISSUE #1327: SMAP: Router actions blocked during tile loading process
### Steps to reproduce the error
- Go to a project with lots of records
- While "loading locations" is displayed, click on a location and click "Open Location"

### Actual behavior
- continue to wait

### Expected behavior
- Tile loading is a non-blocking async call.

### PR DESCRIPTION
**Pull request by linzjax**
_NOTE: __NOT__ BEING MERGED INTO MASTER. BEING MERGED INTO `map-spa-router`_

#### Ready for this?
- Addresses #1060 and #1059 for everything except Location adding/editing/deleting
- Addresses #1061 by adding `/?coords=` to the URL and introducing L.Hash
  - The only times the map should move of its own accord:
  - [ ] When a link to a project contains coordinates in the URL.
  - [ ] When a link to location/relationship detail page does NOT contain coordinates in the URL.
- Fixes #1207. Handles permission errors are redirects for anonymous and non-permissioned users.
- Fixes #1206. Added `/?tab=` to the URL to allow for better redirecting between tabs and external pages.
- Fixes #1310. Coordinates for selected location is sent through the response header, parsed by the router, and the map the centers on the coordinates regardless of whether or not the geometry is loaded. This forces the TileLoader to load this geometry.
- Fixes #1327. A timeout alert has been set around the router so that it is given priority over recursing TileLoader functions. It's not seamless, but it's not totally blocked.

#### Other things of note
- Introduced router_mixin for all functions that update the state/DOM of the current page.
- Refactored `routes` in `routes.js` to need only two functions
  - The route's controller function takes `updatePage()` which sets the basic page template (detail panel, modal, sidebar, page-title).
  - The route's eventhook function takes `updateState()` which calls all state changes and event hooks required (setting current location/relationship, attaching forms to the ajax interceptor, adding datatablehooks).
  - Both functions take a key/value dictionary has the "kwargs" argument to cut down on the number of different places a function is called. It contains a standard set of rules (i.e. is there a modal? True or False), then the router_mixin determines which functions to call.
- Added hooks to the HTML to allow ajax interception of form submission.
- Added mobile compatibility for displaying and hiding project detail panel
- Sidebar tabs are updated when switching from map to overview.
- Custom headers added to async responses to help parse form/permission errors.

### Bugs to double check from last two PRs:

- [ ] Router actions not (horribly) blocked during tile loading process (#1327)
- [x] Accessing a location or relationship from Url should center on corresponding location geometry (#1310)
  - [x] Still works if location geometry is far outside of project extent.
- [x] Router handles permission errors for both Anonymous and non-permissioned users. (Test accessing location through external link) (#1207).
- [x] Location detail tabs should maintain state. Things to check:
  - [x] Accessing a resource and pressing the back button should return you to the resource tab.
  - [x] Accessing a relationship and pressing back should take you to the relationship tab.
  - [x] Switching between locations, new location should default to overview.
  - [x] Accessing a location with specific tab through url should automatically display that tab.
- [x] When first accessing a project url with coordinates, the map zooms to those coordinates (check project overview, project map, location detail pages, and relationship detail pages)
- [x] DataTables is formatting correctly
- [ ] Forms with conditional attributes work, and required fields are starred
- [x] Map doesn't move after deleting a location
- [x] Map doesn't move after canceling 'add location'
- [x] When processing permissions/errors, router checks header of response, not the response message.

### When should this PR be merged
- Whenever.

### Risks
- Even MORE code than the last two PRs 👏 
- There were several things that got lost in rebase transition since things were moved from `view/default.py` to `views/async.py` and so much of the templates ended up on the cutting room floor. I think I've caught everything, but don't assume it's a "feature" if something is missing. Chances are it's a 🐛 

### Follow-up actions
- Once this goes into `map-spa-router`, it should be good to go into `map-spa`. I believe @bjohare's branch is off of `map-spa-router`, which is why I want that one to get caught up first.
- Said it before, and I'll say it again, once we start looking at merging into master, we're gonna need to go over this with a fine-tooth comb to make sure nothing got dropped and we don't have any weird compatibility issues.
- I'll squash everything into one ridiculous long PR once it's been approved, but I was scared to lose the history before everything's been approved.
- My drinking lots of 🍻 

---

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
- [ ] **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
- [ ] **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?** 
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: cadasta/core/static/js/smap/router.js [hunk 1]
```diff
@@ -1,62 +1,87 @@
...
+          }
+
+          if (permission_error) {
+            if (rm.getLastHashPath() && rm.getLastHashPath() !== '/' && rm.getLastHashPath() !== '/overview') {
+              history.back();
+            } else if (!window.location.hash.includes('/overview')) {
+              window.location.hash = "/overview";
+            }
+            if ($('.alert-warning').length === 0) {
+              console.log('is it alert warning?');
```
Comment (oliverroick) [inline cadasta/core/static/js/smap/router.js]: It is, you can remove the log statement.

**Comment by linzjax** [comment]
> Opening a location from a marker does not work, maybe the onclick event handler isn't passed to the marker.

Oh weird yeah. It's not getting passed to L.deflate... I'll look at that.

#### Optional:
> The URL schema is a bit weird. Instead of ?tab=overview/?coords=14/51.4796/-0.1700 it should be ?tab=overview&coords=14/51.4796/-0.1700. But that's a minor issue; we should ignore it when it's too difficult to fix at this point.

This shouldn't be too heinous... I'll take a look.

> When I'm on the relationship tab, then click on one of the parties, and then from the party detail page use the browser's back button the map zooms to the previous location, but the current location is only highlighted when I click on one of the tabs in the location detail page. (It sometimes works, sometimes not, can't figure out when though. Maybe it has to do with the complex London dataset I'm testing with, it takes long to load and to render, and it's not very realistic. So it might not be a problem when we have real data.)

This is that weird gap between the end of TileLoad requests, and then actually adding of geometries to the map. I'm not sure how to track when the geometry is actually added to the page.

> When I paste the link to a location, and the link doesn't contain coordinates and the URL ends with a /, there's an error: Uncaught TypeError: Cannot read property 'el' of undefined at router.js:43.

Aaah dang I should fool proof that. That shouldn't be to bad to fix.

File: cadasta/core/static/js/smap/router_mixins.js [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+  uploadResourceHooks: function() {
+    this.addEventScript('file-upload');
+  },
+
+  addResourceHooks: function() {
+    this.addEventScript('script_add_lib');
+  },
+
+  relationshipHooks: function() {
+    this.addEventScript('rel_new_item');
```
Comment (bjohare) [inline cadasta/core/static/js/smap/router_mixins.js]: Minor issue but I get a `TypeError` here when trying to add a relationship. If the form is invalid and I try to save it the error is thrown.
eventhooks

**Comment by linzjax** [comment]
@oliverroick 
- I followed Katrina's suggestion and used the catiline library to turn part of the tileloader into a webworker. It sped up the process, still not perfect, but at least it doesn't contiain ranom timeouts now.
- No longer using deflate, so popup issue shouldn't be a problem.
- updated the url patter to match convetion.
- RE location coloring: I'm ignoring this for now. I'll open a ticket to deal with it later.
- Fool proofed the hash patter while updating the url pattern.

@bjohare 
- Ah, good catch. Turns out there were a couple of things broken things in the forms. Should be fixed now.

**Pull request merged**